src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs

```diff
@@ -0,0 +1,87 @@
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.IO;
+using System.Reflection;
+using System.Xml;
+namespace MSBuild
+{
+    public class ValidateMSBuildPackageDependencyVersions : Task
+    {
+        [Required]
+        public string AppConfig { get; set; }
+        [Required]
+        public string AssemblyPath { get; set; }
+
+        public override bool Execute()
+        {
+            XmlDocument doc = new XmlDocument();
+            doc.Load(AppConfig);
+            foreach (var topLevelElement in doc.ChildNodes)
+            {
+                if (topLevelElement is XmlElement topLevelXmlElement && topLevelXmlElement.Name.Equals("configuration", StringComparison.OrdinalIgnoreCase))
+                {
+                    foreach (var configurationElement in topLevelXmlElement.ChildNodes)
+                    {
+                        if (configurationElement is XmlElement configurationXmlElement && configurationXmlElement.Name.Equals("runtime", StringComparison.OrdinalIgnoreCase))
+                        {
+                            foreach (var runtimeElement in configurationXmlElement.ChildNodes)
+                            {
+                                if (runtimeElement is XmlElement runtimeXmlElement && runtimeXmlElement.Name.Equals("assemblyBinding", StringComparison.OrdinalIgnoreCase))
+                                {
+                                    foreach (var assemblyBindingElement in runtimeXmlElement.ChildNodes)
+                                    {
+                                        if (assemblyBindingElement is XmlElement assemblyBindingXmlElement && assemblyBindingXmlElement.Name.Equals("dependentAssembly", StringComparison.OrdinalIgnoreCase))
+                                        {
+                                            string name = string.Empty;
+                                            string version = string.Empty;
+                                            foreach (var dependentAssembly in assemblyBindingXmlElement.ChildNodes)
+                                            {
```

Comment:
You can simplify this a lot using [`doc.SelectNodes(xpath)`](https://docs.microsoft.com/dotnet/standard/data/xml/select-nodes-using-xpath-navigation).

Reply:
This seems surprisingly unreliable in my testing. I tried doc.SelectNodes("dependentAssembly[assemblyIdentity][bindingRedirect]"), and it got no results, so I'm assuming it only looks at children of its direct parent.

Then I tried doc.SelectSingleNode("configuration").SelectSingleNode("runtime").SelectSingleNode("assemblyBinding"), and the first two SelectSingleNode calls returned what I'd expected, and the third returned null. I tried changing the capitalization plus a couple other things, but no dice.

I did manage to get it working with some sql statements, but it looks a little odd. Still a good bit cleaner.

Reply:
That sounds like XName shenanigans to me. You might need to try a fully qualified xname for the elements you're querying for in order for the nice APIs to do what you expect. This bites me _every time_ I use the XML APIs.

Reply:
Like adding a namespace or adding a list of the parents in order? Namespaces seem to be a thing for XmlElements, but I don't know what it would be for arbitrary elements in app.configs. Listing parents is doable, but I don't see anywhere that suggests that's what it would be.

Reply:
Oh, wait. It lists something afterwards. Let me try that.

Reply:
That does seem to be the namespace, but I haven't figured out how to incorporate it into the name such that it finds it.

Reply:
```c#
XmlNamespaceManager namespaceManager = new XmlNamespaceManager(doc.NameTable);
namespaceManager.AddNamespace("asm", "urn:schemas-microsoft-com:asm.v1");

var dependentAssemblyElements = doc.DocumentElement.SelectNodes("/configuration/runtime/asm:assemblyBinding/asm:dependentAssembly", namespaceManager);
```

Reply:
(this is so much easier with XDocument that I'm starting to rethink the suggestion I gave to use System.Xml)

Reply:
I don't think I would've guessed that, and it wasn't at all clear where the namespaces were coming from in the examples I found. Thanks!

I think it's reasonably clean now, albeit with a few long/ugly lines.